### PR TITLE
Add GitHub Actions guide

### PR DESCRIPTION
**Pull request by usha-mandya**
Add docs on how to configure GitHub Actions

### CODE REVIEW

**Comment by usha-mandya** [comment]
@StefanScherer @nebuk89 Could you PTAL?

File: ci-cd/github-actions.md [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+
+## Set up a Docker project
+
+Let’s get started. This guide uses a simple Docker project as an example. The [SimpleWhaleDemo](https://github.com/usha-mandya/SimpleWhaleDemo){:target="_blank" rel="noopener" class="_"} repository contains an Ngnix alpine image. You can either clone this repository, or use your own Docker project.
+
+![SimpleWhaleDemo](images/simplewhaledemo.png){:width="500px"}
+
+Before we start, ensure you can access [Docker Hub](https://hub.docker.com/) from any workflows you create. To do this:
+
+1. Add your Docker ID and Personal Access Token (PAT) as secrets to GitHub.
```
Comment (StefanScherer) [inline ci-cd/github-actions.md:24]:
```suggestion
1. Add your Docker ID as a secret to GitHub.
```
Navigate to your GitHub repository and click **Settings > Secrets**.

github-new-secret

Create a new secret with the name `DOCKER_HUB_USERNAME` and with your Docker ID as value.

File: ci-cd/github-actions.md [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+2. To create a new token, go to [Docker Hub Settings](https://hub.docker.com/settings/security) and then click **New Access Token**.
+3. Let’s call this token **simplewhaleci**.
+
+    ![New access token](images/github-access-token.png){:width="500px"}
+
+4. Now, add this token to your username as secrets into the GitHub secrets UI.
```
Comment (StefanScherer) [inline ci-cd/github-actions.md:30]:
```suggestion
4. Now, add this Personal Access Token (PAT) as a second secret into the GitHub secrets UI with the name `DOCKER_HUB_ACCESS_TOKEN`.
```

Comment (StefanScherer) [reply ci-cd/github-actions.md:30]: I suggest to write the secret names in the docs to make it readable in addition to the screenshots.

File: ci-cd/github-actions.md [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+jobs:
+
+  build:
+    runs-on: ubuntu-latest
+```
+
+Now, we can add the steps required. The first one is to use our PAT and username to log into Docker Hub. The second is the Builder, the action  uses BuildKit under the hood through a simple Buildx action which we will also setup
+
+```yaml
+    Steps:
```
Comment (StefanScherer) [inline ci-cd/github-actions.md:74]:
```suggestion
    steps:
```

lower case

File: ci-cd/github-actions.md [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+
+
+      - name: Login to Docker Hub
+        uses: docker/login-action@v1
+        with:
+          username: ${{ secrets.DOCKER_HUB_USERNAME }}
```
Comment (StefanScherer) [inline ci-cd/github-actions.md:80]: Oh, these double curly braces here and further down in the code samples get lost in Jekyll. 😢 

docs-yaml-vars-missing

Looks like there are some workarounds to escape the curly braces https://stackoverflow.com/questions/24102498/escaping-double-curly-braces-inside-a-markdown-code-block-in-jekyll

File: ci-cd/github-actions.md [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+
+```yaml
+
+      - name: Cache Docker layers
+        uses: actions/cache@v2
+        with:
+          path: /tmp/.buildx-cache
+          key: ${{ runner.os }}-buildx-${{ github.sha }}
+          restore-keys: |
+            ${{ runner.os }}-buildx-
```
Comment (StefanScherer) [inline ci-cd/github-actions.md:121]: The `restore-keys` is different to `key`?
I just want to double check if this is correct. 😅 

Seems correct as it is used in https://github.com/metcalfc/docker-action-examples/blob/main/.github/workflows/release.yml#L44-L50 👍

File: ci-cd/github-actions.md [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+
+## Conclusion
+
+In this guide, you have learnt how to set up GitHub Actions workflow to an existing Docker project, optimize your workflow to improve build times and reduce the number of pull requests, and finally, we learnt how to push only specific versions to Docker Hub.
+
+## Next steps
+
+You can now consider setting up nightly tests against the latest tag, test each PR, or do something more elegant with the tags we are using and make use of the Git tag for the same tag in our image.
+
+To look at how you can do one of these, or to get a full example on how to set up what we have accomplished today, check out Chad’s repo which runs you through this and more details on our latest GitHub action - https://github.com/metcalfc/docker-action-examples.
```
Comment (StefanScherer) [inline ci-cd/github-actions.md:218]: Do we want to make it a link?

```suggestion
To look at how you can do one of these, or to get a full example on how to set up what we have accomplished today, check out Chad’s repo which runs you through this and more details on our latest GitHub action - [https://github.com/metcalfc/docker-action-examples](https://github.com/metcalfc/docker-action-examples).
```

**Review by StefanScherer — CHANGES_REQUESTED** [review]
Thanks @usha-mandya 
I've added a few suggestions and we have a problem rendering the yaml at the moment. Jekyll needs some help to show the variables properly.

File: ci-cd/github-actions.md (hunk 3 above)
Comment (usha-mandya) [reply ci-cd/github-actions.md]: Fixed

File: ci-cd/github-actions.md (hunk 6 above)
Comment (usha-mandya) [reply ci-cd/github-actions.md]: Updated

File: ci-cd/github-actions.md (hunk 4 above)
Comment (usha-mandya) [reply ci-cd/github-actions.md:85]: I've added `{% raw %}` and `{% endraw %}` tags

Comment (usha-mandya) [reply ci-cd/github-actions.md:85]: image

File: ci-cd/github-actions.md [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+## Optimizing the workflow
+
+Next, let’s look at how we can optimize the GitHub Actions workflow through build cache. This has two main advantages:
+
+1. Build cache reduces the build time as it will not have to re-download all of the images, and
+2. It also reduces the number of pulls we complete against Docker Hub. We need to make use of GitHub cache to make use of this.
+
+Let us set up a Builder with a build cache. First, we need to set up cache for the builder.  In this example, let us add the path and keys to store this under using GitHub cache for this.
+
+```yaml
```
Comment (StefanScherer) [inline ci-cd/github-actions.md:119]: This needs another `{% raw %} ... {% endraw %}` block.

File: ci-cd/github-actions.md [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+
+Now, let’s set up a second GitHub action file to store our latest commit as an image in the GitHub registry. You may want to do this to:
+
+1. Run your nightly tests or recurring tests, or
+2. To share work in progress images with colleagues.
+
+Let’s clone our previous GitHub action and add back in our previous logic for all pushes. This will mean we have two workflow files, our previous one and our new one we will now work on.
+Next, change your Docker Hub login to a GitHub container registry login:
+
+```yaml
```
Comment (StefanScherer) [inline ci-cd/github-actions.md:199]: This needs another `{% raw %} ... {% endraw %}` block.

File: ci-cd/github-actions.md [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+        uses: docker/login-action@v1
+        with:
+        registry: ghcr.io
+          username: ${{ github.repository_owner }}
+          password: ${{ secrets.GHCR_TOKEN }}
+```
+
+Remember to change how the image is tagged. The following example keeps ‘latest’ as the only tag. However, you can add any logic to this if you prefer:
+
+```yaml
```
Comment (StefanScherer) [inline ci-cd/github-actions.md:210]: This needs another `{% raw %} ... {% endraw %}` block.

**Review by StefanScherer — CHANGES_REQUESTED** [review]
The steps how to create the secrets is much clearer now!

A few yaml blocks need the raw/endraw block to be displayed correctly.

**Review by StefanScherer — APPROVED** [review]
LGTM

**Pull request merged**